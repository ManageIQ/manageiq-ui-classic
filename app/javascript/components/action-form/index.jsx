import React, { useState, useEffect, useMemo } from 'react';
import MiqFormRenderer from '@@ddf';
import PropTypes from 'prop-types';
import createSchema from './action-form.schema';
import miqRedirectBack from '../../helpers/miq-redirect-back';
import { dataHelper, constructAeHash } from './helper';

const ActionForm = ({
  recordId, inheritTags, availableAlerts, tags, ansibleInventory, snapshotAge, parentType, inventoryType,
}) => {
  const [{ initialValues, isLoading }, setState] = useState({
    isLoading: !!recordId,
  });

  const submitLabel = !!recordId ? __('Save') : __('Add');
  const promise = useMemo(() => API.options('/api/actions'), []);

  const onSubmit = (values) => {
    miqSparkleOn();
    const data = dataHelper(values);

    const request = recordId ? API.patch(`/api/actions/${recordId}`, data) : API.post('/api/actions', data);
    request.then(() => {
      const message = sprintf(
        recordId
          ? __('New action "%s" has been successfully queued.')
          : __('Add of action "%s" has been successfully queued.'),
        data.name,
      );
      miqRedirectBack(message, 'success', '/miq_action/show_list');
    }).catch(miqSparkleOff);
  };

  const onCancel = () => {
    const message = sprintf(
      recordId
        ? __('Edit of Action "%s" was canceled by the user.')
        : __('Creation of new Action was canceled by the user.'),
      initialValues && initialValues.name,
    );
    miqRedirectBack(message, 'warning', '/miq_action/show_list');
  };

  useEffect(() => {
    if (recordId) {
      API.get(`/api/actions/${recordId}`).then((initialValues) => {
        if (initialValues.action_type === 'custom_automation') {
          initialValues.options.ae_hash = constructAeHash(initialValues.options.ae_hash);
          setState({ initialValues, isLoading: false });
        } else if (initialValues.action_type === 'tag') {
          // eslint-disable-next-line no-useless-escape
          initialValues.options.tags = initialValues.options.tags[0].replace('\/managed\/', '');
          setState({ initialValues, isLoading: false });
        } else {
          setState({ initialValues, isLoading: false });
        }
      });
    }
  }, [recordId]);

  return !isLoading && (
    <MiqFormRenderer
      schema={createSchema(!!recordId, promise, inheritTags, availableAlerts, tags, ansibleInventory, snapshotAge, parentType, inventoryType)}
      initialValues={initialValues}
      canReset={!!recordId}
      onSubmit={onSubmit}
      onCancel={onCancel}
      buttonsLabels={{ submitLabel }}
    />
  );
};

ActionForm.propTypes = {
  inheritTags: PropTypes.arrayOf(PropTypes.any).isRequired,
  availableAlerts: PropTypes.arrayOf(PropTypes.any).isRequired,
  recordId: PropTypes.string,
  tags: PropTypes.objectOf(PropTypes.any).isRequired,
  ansibleInventory: PropTypes.arrayOf(PropTypes.any).isRequired,
  snapshotAge: PropTypes.arrayOf(PropTypes.any).isRequired,
  parentType: PropTypes.arrayOf(PropTypes.any).isRequired,
  inventoryType: PropTypes.string,
};

ActionForm.defaultProps = {
  recordId: undefined,
  inventoryType: undefined,
};

export default ActionForm;
