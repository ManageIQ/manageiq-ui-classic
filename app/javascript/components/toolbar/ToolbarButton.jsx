import React, { useContext } from 'react';
import PropTypes from 'prop-types';

import { adjustColor, isEnabled } from './utility';
import CountContext from './ToolbarContext';

const classNames = require('classnames');

const ButtonIcon = ({
  img_url: imgUrl, icon, color, enabled,
}) => {
  if (icon) {
    return <i className={icon} style={{ color: adjustColor(color, enabled) }} />;
  }

  if (imgUrl && !icon) {
    return <img alt="Button icon" src={imgUrl} />;
  }

  return null;
};

ButtonIcon.propTypes = {
  img_url: PropTypes.string,
  icon: PropTypes.string,
  color: PropTypes.string,
  enabled: PropTypes.bool.isRequired,
};

ButtonIcon.defaultProps = {
  img_url: null,
  icon: null,
  color: null,
};

export const ToolbarButton = (props) => {
  const count = useContext(CountContext);
  const {
    onwhen, enabled, id, name, title, text, selected,
  } = props;
  const disabled = !(onwhen ? isEnabled(onwhen, count) : enabled);

  return (
    <button
      type="button"
      id={id}
      name={name}
      title={title}
      disabled={disabled}
      className={classNames('btn btn-default toolbar-button', { active: selected, disabled })}
      onClick={() => props.onClick(props)}
    >
      { ButtonIcon(props) }
      { text }
    </button>
  );
};

ToolbarButton.propTypes = {
  title: PropTypes.string,
  id: PropTypes.string,
  name: PropTypes.string,
  text: PropTypes.string,
  selected: PropTypes.bool,
  enabled: PropTypes.bool.isRequired,
  onwhen: PropTypes.string,
  onClick: PropTypes.func.isRequired,
};

ToolbarButton.defaultProps = {
  title: null,
  id: null,
  name: null,
  text: null,
  onwhen: null,
  selected: null,
};
