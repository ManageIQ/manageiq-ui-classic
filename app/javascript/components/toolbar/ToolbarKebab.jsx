import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';
import {
  OverflowMenu,
  SideNavMenu,
  SideNavMenuItem,
  SideNavDivider,
  SideNavLink,
  SideNavItems,
} from 'carbon-components-react';
import { ToolbarClick } from './ToolbarClick';

const KebabListItem = (item, props) => {
  if (item.type === 'separator') {
    return <SideNavDivider key={item.id} />;
  }

  const ButtonIcon = () => (
    <i className={item.icon || ''} style={{ color: item.color || '' }} />
  );

  if (item.type === 'buttonSelect') {
    return (
      <SideNavMenu
        key={item.id}
        renderIcon={ButtonIcon}
        title={item.text || item.title}
      >
        { item.items.filter((i) => !i.hidden).map((i) => (
          <SideNavMenuItem
            key={i.id}
            onClick={props.onClick && i.enabled ? (() => props.onClick(i)) : null}
          >
            <ToolbarClick {...i} />
          </SideNavMenuItem>
        )) }

      </SideNavMenu>
    );
  }

  return (
    <SideNavLink onClick={item.onClick && item.enabled ? (() => item.onClick(item)) : null}><ToolbarClick key={item.id} {...item} /></SideNavLink>
  );
};

KebabListItem.propTypes = {
  item: PropTypes.arrayOf(PropTypes.any),
  onClick: PropTypes.func.isRequired,
};

KebabListItem.defaultProps = {
  item: null,
};

// eslint-disable-next-line no-unused-vars
export const DropDownMenu = forwardRef((props, ref) => (
  <SideNavItems className="button_groups">
    {props.items.map((item) => KebabListItem(item, props))}
  </SideNavItems>
));

DropDownMenu.propTypes = {
  items: PropTypes.arrayOf(PropTypes.any),
};

DropDownMenu.defaultProps = {
  items: null,
};

export const ToolbarKebab = forwardRef((props, ref) => (
  <div className="btn-group kebab">
    <OverflowMenu ref={ref}>
      <DropDownMenu {...props} ref={ref} />
    </OverflowMenu>
  </div>
));

ToolbarKebab.propTypes = {
  id: PropTypes.string.isRequired,
  items: PropTypes.arrayOf(PropTypes.any),
  onClick: PropTypes.func.isRequired,
  icon: PropTypes.string,
  color: PropTypes.string,
  text: PropTypes.string,
  title: PropTypes.string,
};

ToolbarKebab.defaultProps = {
  color: null,
  text: null,
  icon: null,
  title: null,
  items: null,
};
