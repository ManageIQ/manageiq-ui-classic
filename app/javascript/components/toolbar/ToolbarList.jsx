import React, { useContext, useState } from 'react';
import PropTypes from 'prop-types';
import { OverflowMenu, OverflowMenuItem } from 'carbon-components-react';
import { ChevronDown20 } from '@carbon/icons-react';
import { MenuIcon } from './MenuIcon';
import { isEnabled } from './utility';
import { ToolbarClick } from './ToolbarClick';
import CountContext from './ToolbarContext';

export const ToolbarList = (props) => {
  const count = useContext(CountContext);
  const { items, title, id } = props;
  // Set this true for overflowmenu keydown event
  const [overflowTab, setOverflowTab] = useState(false);

  // Filter out invisible items.
  const visibleItems = items.filter((i) => !i.hidden);

  // Do not render at all if no child is visible.
  if (visibleItems.length === 0) {
    return null;
  }

  // Calculate item's enable state based on item's initial enable state, onwhen and count.
  // Toolbar is disabled if no item is enabled.
  let isToolbarEnabled = false;
  const enabledItems = visibleItems.map((i) => {
    const enabled = i.onwhen ? isEnabled(i.onwhen, count) : i.enabled;
    isToolbarEnabled = isToolbarEnabled || enabled;
    return {
      ...i,
      enabled,
    };
  });

  const keydownFunc = (event) => {
    if (event.keyCode === 9 || event.keyCode === 27) {
      setOverflowTab(true);
    }
  };

  const closeFunc = () => {
    if (overflowTab === true) {
      document.getElementById(props.id).focus();
      setOverflowTab(false);
    }
  };

  return (
    <OverflowMenu
      ariaLabel={title}
      id={id}
      floatingmenu="true"
      title={title}
      iconDescription=""
      className={!isToolbarEnabled ? 'overflow-menu-disabled' : ''}
      disabled={!isToolbarEnabled}
      onClose={closeFunc}
      renderIcon={() => (
        <div className="toolbar-overflow">
          { props.icon && <i className={props.icon} style={{ color: props.color }} /> }
          <ChevronDown20 />
          {' '}
          <span>{ props.text ? props.text : (props.title && props.title)}</span>
        </div>
      )}
    >
      { enabledItems.map((item) => (
        <OverflowMenuItem
          className="overflow-options"
          key={item.id}
          aria-label={item.title}
          hasDivider={(item.type === 'separator')}
          itemText={<MenuIcon text={item.text ? item.text : (item.title && item.title)} icon={item.icon} color={item.color} />}
          disabled={!item.enabled}
          onKeyDown={keydownFunc}
          title={item.title}
          requireTitle
          onClick={props.onClick && item.enabled ? (() => props.onClick(item)) : null}
        >
          <ToolbarClick {...props} />
        </OverflowMenuItem>

      )) }
    </OverflowMenu>
  );
};

ToolbarList.propTypes = {
  id: PropTypes.string.isRequired,
  items: PropTypes.arrayOf(PropTypes.any),
  onClick: PropTypes.func.isRequired,
  icon: PropTypes.string,
  color: PropTypes.string,
  text: PropTypes.string,
  title: PropTypes.string,
};

ToolbarList.defaultProps = {
  color: null,
  text: null,
  icon: null,
  title: null,
  items: null,
};
