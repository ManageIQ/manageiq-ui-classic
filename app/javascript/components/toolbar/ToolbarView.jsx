import React from 'react';
import PropTypes from 'prop-types';
import { adjustColor } from './utility';

const classNames = require('classnames');

export const ToolbarView = ({ views, onClick }) => (
  <div className="toolbar-pf-action-right">
    <div className="form-group toolbar-pf-view-selector">
      {views.map(view => (
        <button
          key={view.id}
          id={view.id}
          name={view.name}
          title={view.title}
          className={classNames('btn btn-link', { active: view.selected })}
          data-url={view.url}
          data-url_parms={view.url_parms}
          data-send_checked={view.send_checked ? 'true' : ''}
          data-prompt={view.prompt}
          data-popup={view.popup}
          disabled={!view.enabled}
          onClick={() => onClick(view)}
        >
          <i className={view.icon} style={{ color: adjustColor(view.color, view.enabled) }} />
        </button>
        ))
      }
    </div>
  </div>
);

ToolbarView.propTypes = {
  views: PropTypes.arrayOf(PropTypes.any),
  onClick: PropTypes.func.isRequired,
};

