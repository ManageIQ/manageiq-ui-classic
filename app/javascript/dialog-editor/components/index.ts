import box from './box';
import dialogEditor from './dialog-editor';
import field from './field';
import modal from './modal';
import modalBox from './modal-box';
import modalField from './modal-field';
import modalFieldTemplate from './modal-field-template';
import modalTab from './modal-tab';
import tabList from './tab-list';
import toolbox from './toolbox';
import treeSelector from './tree-selector';
import validation from './validation';

export default (module: ng.IModule) => {
  box(module);
  dialogEditor(module);
  field(module);
  modal(module);
  modalBox(module);
  modalField(module);
  modalFieldTemplate(module);
  modalTab(module);
  tabList(module);
  toolbox(module);
  treeSelector(module);
  validation(module);
};
