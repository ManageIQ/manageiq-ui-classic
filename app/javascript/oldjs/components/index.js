require('./ae_resolve_options/ae-resolve-options.js');
require('./ansible-raw-stdout.js');
require('./auth-credentials.js');
require('./cloud_volume_backup/cloud-volume-backup-form.js');
require('./datetime-delay-picker.js');
require('./generic_object/assign-buttons.js');
require('./generic_object/custom-button-group-form.js');
require('./generic_object/custom-image.js');
require('./generic_object/generic-object-definition-toolbar.js');
require('./generic_object/main-custom-button-form.js');
require('./generic_object/main-custom-button-group-form.js');
require('./index.js');
require('./miq-button.js');
require('./miq-request-options.js');
require('./miq_ae_class/namespace-form.js');
require('./ops/log-collection-form.js');
require('./pf_charts/aggregate-status-card.component.js');
require('./pf_charts/c3-chart.component.js');
require('./pf_charts/empty-chart.component.js');
require('./pf_charts/heatmap-legend.component.js');
require('./pf_charts/heatmap.component.js');
require('./provider-option-field-input.js');
require('./provier-option-section.js');
require('./sanitize.js');
require('./verify-button.js');
require('./vm_cloud/vm-cloud-add-security-group.js');
require('./vm_cloud/vm-cloud-remove-security-group.js');
require('./vm_cloud/vm_cloud_evacuate_form.js');
