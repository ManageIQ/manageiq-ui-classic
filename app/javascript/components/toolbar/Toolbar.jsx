/* eslint-disable react/destructuring-assignment */
import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';
import { ToolbarView } from './ToolbarView';
import { ToolbarButton } from './ToolbarButton';
import { ToolbarList } from './ToolbarList';
import { ToolbarKebab } from './ToolbarKebab';
import CountContext from './ToolbarContext';

export const ButtonType = {
  BUTTON: 'button',
  BUTTON_TWO_STATE: 'buttonTwoState',
  BUTTON_SELECT: 'buttonSelect',
  SEPARATOR: 'separator',
  KEBAB: 'kebab',
};
const ref = React.createRef();
const isButton = (item) => (item.type === ButtonType.BUTTON);
const isButtonTwoState = (item) => (item.type === ButtonType.BUTTON_TWO_STATE);
const isButtonSelect = (item) => (item.type === ButtonType.BUTTON_SELECT);
const isKebabMenu = (item) => (item.type === ButtonType.KEBAB);

const isButtonSelectVisible = (props) => props.items.filter((i) => !i.hidden).length !== 0;

const isVisibleButtonOrSelect = (item) => (
  item.type && (
    (isButtonSelect(item) && isButtonSelectVisible(item))
    || (isButton(item) && !item.hidden)
    || (isButtonTwoState(item) && !item.hidden)
    || (isKebabMenu(item) && !item.hidden)
  )
);

const toolbarGroupHasContent = (group) =>
  group
    && group.filter((item) => item
      && isVisibleButtonOrSelect(item)).length !== 0;

export const ButtonCase = forwardRef(({ item, index, onClick }, ref) => {
  if (isButton(item) || isButtonTwoState(item)) {
    return <ToolbarButton key={index} {...item} onClick={onClick} />;
  } if (isButtonSelect(item) && (item.items.length > 0)) {
    return <ToolbarList key={index} {...item} onClick={onClick} />;
  }
  if (isKebabMenu(item) && (item.items.length > 0)) {
    return <ToolbarKebab key={index} {...item} onClick={onClick} ref={ref} />;
  }
  return null;
});

ButtonCase.propTypes = {
  item: PropTypes.objectOf(PropTypes.any).isRequired,
  index: PropTypes.number,
  onClick: PropTypes.func.isRequired,
};

ButtonCase.defaultProps = {
  index: 0,
};

/* custom buttons have ID's starting with this: */
const CUSTOM_ID = 'custom_';

const collapseOverlimit = (itemsGroup, kebabLimit) => {
  let numItems = 0;

  return itemsGroup.reduce((acc, i) => {
    if (i.id.includes(CUSTOM_ID)) {
      if (numItems >= kebabLimit) {
        acc[acc.length - 1].items.push(i);
      } else {
        numItems += 1;
        acc.splice(acc.length - 1, 0, i);
      }
    } else {
      acc.splice(acc.length - 1, 0, i);
    }
    return acc;
  }, [{ type: ButtonType.KEBAB, items: [] }]);
};

/*
 * kebabLimit === -1 means no compacting
 * kebabLimit === 0 means always compact into kebab
 * other values of kebabLimit give number of items to keep, before the rest is kebabized
*/
const collapseCustomGroups = (itemsGroup, kebabLimit) => (
  (kebabLimit === -1) || (itemsGroup.length < kebabLimit)
    ? itemsGroup
    : collapseOverlimit(itemsGroup, kebabLimit)
);

export const ToolbarGroup = forwardRef(({ group, onClick }, ref) => {
  const visibleItems = group.filter(isVisibleButtonOrSelect);

  if (visibleItems.length === 0) {
    return null;
  }

  return (
    <div className="miq-toolbar-group form-group">
      {visibleItems.map((i, index) => <ButtonCase item={i} key={index} index={index} onClick={onClick} ref={ref} />)}
    </div>
  );
});

ToolbarGroup.propTypes = {
  group: PropTypes.arrayOf(PropTypes.any).isRequired,
  onClick: PropTypes.func.isRequired,
};

export const Toolbar = (props) => (
  <CountContext.Provider value={props.count}>
    <div className="toolbar-pf-actions miq-toolbar-actions">
      { props.groups
        .filter(toolbarGroupHasContent)
        .map((group, index) =>
          // eslint-disable-next-line react/no-array-index-key
          <ToolbarGroup key={index} ref={ref} onClick={props.onClick} group={collapseCustomGroups(group, props.kebabLimit)} />)}
      <ToolbarView onClick={props.onViewClick} views={props.views} />
    </div>
  </CountContext.Provider>
);

Toolbar.propTypes = {
  count: PropTypes.number,
  kebabLimit: PropTypes.number,
  groups: PropTypes.arrayOf(PropTypes.any).isRequired, // array of arrays of buttons
  views: PropTypes.arrayOf(PropTypes.any).isRequired, // array of view buttons
  onClick: PropTypes.func.isRequired,
  onViewClick: PropTypes.func.isRequired,
};

Toolbar.defaultProps = {
  count: 0,
  kebabLimit: 3,
};
