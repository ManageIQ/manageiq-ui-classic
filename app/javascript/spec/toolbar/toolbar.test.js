import React from 'react';
import { mount, shallow } from 'enzyme';

import { Toolbar } from '../../components/toolbar';
import { ToolbarButton } from '../../components/toolbar/ToolbarButton';
import { ToolbarGroup } from '../../components/toolbar/Toolbar';
import { ToolbarList } from '../../components/toolbar/ToolbarList';
import { ToolbarView } from '../../components/toolbar/ToolbarView';
import { ToolbarKebab } from '../../components/toolbar/ToolbarKebab';

import toolbarData from './data/toolbar-big.json';
import viewData from './data/toolbar-view.json';

describe('Toolbar', () => {
  it('Well it works ;-)', () => {
    const t = shallow(<Toolbar
      onClick={() => {}}
      onViewClick={() => {}}
      groups={toolbarData}
      views={viewData}
      count={0}
    />);
    expect(t).toMatchSnapshot();
  });

  it('renders the ToolbarView component', () => {
    const t = shallow(<Toolbar
      onClick={() => {}}
      onViewClick={() => {}}
      groups={toolbarData}
      views={viewData}
      count={0}
    />);
    expect(t.find(ToolbarView)).toHaveLength(1);
  });

  it('renders view buttons', () => {
    const t = mount(<Toolbar
      onClick={() => {}}
      onViewClick={() => {}}
      groups={toolbarData}
      views={viewData}
      count={0}
    />);
    expect(t.find('button.btn.btn-link')).toHaveLength(3);
  });

  it('renders groups', () => {
    const t = shallow(<Toolbar
      onClick={() => {}}
      onViewClick={() => {}}
      groups={toolbarData}
      views={viewData}
      count={0}
    />);
    expect(t.find(ToolbarGroup)).toHaveLength(4);
  });

  it('renders simple buttons', () => {
    const t = mount(<Toolbar
      onClick={() => {}}
      onViewClick={() => {}}
      groups={toolbarData}
      views={viewData}
      count={0}
    />);
    expect(t.find(ToolbarButton)).toHaveLength(2);
    expect(t.find(ToolbarButton).at(0).containsMatchingElement(<i className="fa fa-refresh fa-lg" />))
      .toBeTruthy();
  });

  it('renders drop-down buttons', () => {
    const t = mount(<Toolbar
      kebabLimit={0}
      onClick={() => {}}
      onViewClick={() => {}}
      groups={toolbarData}
      views={viewData}
      count={0}
    />);
    expect(t.find(ToolbarList)).toHaveLength(7);
    expect(t.find(ToolbarList).at(1).text()).toContain('Configuration');
    expect(t.find(ToolbarList).at(1).containsMatchingElement(<i className="fa fa-cog fa-lg" />))
      .toBeTruthy();
  });

  it('renders applies kebabLimit properly', () => {
    const t = mount(<Toolbar
      kebabLimit={3}
      onClick={() => {}}
      onViewClick={() => {}}
      groups={toolbarData}
      views={viewData}
      count={0}
    />);
    expect(t.find(ToolbarList)).toHaveLength(10);
  });

  it('renders kebab buttons', () => {
    const t = mount(<Toolbar
      onClick={() => {}}
      onViewClick={() => {}}
      groups={toolbarData}
      views={viewData}
      count={0}
    />);
    expect(t.find(ToolbarKebab)).toHaveLength(1);
  });
});
