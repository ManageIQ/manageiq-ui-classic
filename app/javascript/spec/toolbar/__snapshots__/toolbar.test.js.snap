// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toolbar Well it works ;-) 1`] = `ShallowWrapper {}`;
